subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy

```diff
@@ -435,6 +441,101 @@ class MissingTaskDependenciesIntegrationTest extends AbstractIntegrationSpec imp
         executedAndNotSkipped(":fooReport", ":barReport")
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/20391")
+    def "running tasks in parallel with exclusions does not cause incorrect builds"() {
```

Comment:
This reproducer is clearly inspired by the problem we found in our own build. It is possible to deduce this by looking at the issue, then following the link to a PR fixing the problem in our own build mentioned in the description of the issue. I think we could also mention this in a comment here, and that would make it most understandable when someone bumps into this later.

Reply:
I added a comment.